pkg/dogstatsd/listeners/packet_buffer.go

```diff
@@ -0,0 +1,61 @@
+package listeners
+
+import (
+	"sync"
+	"time"
+)
+
+type packetBuffer struct {
+	packets       Packets
+	flushTimer    *time.Ticker
+	bufferSize    uint
+	outputChannel chan Packets
+	closeChannel  chan struct{}
+	m             sync.Mutex
+}
+
+func newPacketBuffer(bufferSize uint, flushTimer time.Duration, outputChannel chan Packets) *packetBuffer {
+	pb := &packetBuffer{
+		bufferSize:    bufferSize,
+		flushTimer:    time.NewTicker(flushTimer),
+		outputChannel: outputChannel,
+		packets:       make(Packets, 0, bufferSize),
+		closeChannel:  make(chan struct{}),
+	}
+	go pb.flushLoop()
+	return pb
+}
+
+func (pb *packetBuffer) flushLoop() {
+	for {
+		select {
+		case <-pb.flushTimer.C:
+			pb.m.Lock()
+			pb.flush()
+			pb.m.Unlock()
+		case <-pb.closeChannel:
+			return
+		}
+	}
+}
+
+func (pb *packetBuffer) append(packet *Packet) {
+	pb.m.Lock()
+	defer pb.m.Unlock()
+	if uint(len(pb.packets)) == pb.bufferSize {
```

Comment:
To (potentially) reduce the time to deliver the buffer, I would probably do this the other way around.... First append, then flush if the buffer is full. As opposed to flush if it's full then append, makes no big difference, but the former just provide for an expedited delivery.